sqlgen/db.go

```diff
@@ -100,35 +110,82 @@ func NewDB(conn *sql.DB, schema *Schema) *DB {
 // pairs.
 func (db *DB) WithShardLimit(shardLimit Filter) (*DB, error) {
 	if db.shardLimit != nil {
-		return nil, errors.New("already limited")
+		return nil, errors.New("already has shard limit")
 	}
 
 	dbCopy := *db
 	dbCopy.shardLimit = shardLimit
 	return &dbCopy, nil
 }
 
-func (db *DB) checkFilterAgainstShardLimit(filter Filter) error {
-	if db.shardLimit == nil {
-		return nil
+// WithDynamicLimit is similar to WithShardLimit except that it supports
+// dynamically computing what filter to enforce as a limit, based on a user-
+// provide callback. It may be used together with a shard limit.
+//
+// WithDynamicLimit takes two arguments, `getLimitFilter`, and `onError`,
+// both user-provided callbacks.
+//
+// Before making an SQL query, a Thunder DB handle with a dynamic limit
+// will call `getLimitFilter` with the request context and the name of
+// the table being queried. It will use the sqlgen filter returned by
+// the function as the limit to check for against the current query.
+// The underlying limit check is the same as that used in `WithShardLimit`.
+//
+// If the limit check fails, Thunder will call `shouldContinueOnError` with
+// the resulting error object.  If `shouldContinueOnError` itself returns
+// an error, the query will fail. Otherwise, if the return value is nil Thunder
+// will still carry out the query.
+func (db *DB) WithDynamicLimit(getLimitFilter DynamicLimitFilterCallback, shouldContinueOnError DynamicLimitErrorCallback) (*DB, error) {
```

Comment:
Might be cleaner to export the DynamicLimit struct and taking a DynamicLimit here instead of the 2 callbacks. 

The documentation for the callback functionality can then go in the DynamicLimit doc rather than the WithDynamicLimit doc.

Reply:
Yes, I agree this would be cleaner.

Reply:
i like that passing the functions directly is unambiguous and you cannot mess up.

(i.e. harder to accidentally not pass one of the functions)

Reply:
We could add that check in WithDynamicLimit, that way we could also remove the duplicate checks in checkFilterAgainstLimit(s)

Reply:
sorry i thought about it for another second, i think i'd propose an api that looks like:
```
func ... WithDyanamicLimit(filterCreator DynamicLimitFilterCreator) {
  // ...
}

type DynamicLimitFilterCreator func(ctx context.Context, tableName string) (Filter, bool) # or second return value of type error
```

where a single callback sends both the filter and what to do if it fails. I think that makes the intent clear to the implementer how they should think about the inputs and outputs of the decision they need to write.

Reply:
Ahh, but then how would we notify user code of a filter check failure? We'd like this for people to be able to log failures, etc.

Reply:
We could have `DynamicLimitFilterCreator` return the filter and the error callback instead of a pure error or bool. All the logic is in one place, and we could have internal checks for a nil error callback.

Reply:
We're sticking with 

```
Might be cleaner to export the DynamicLimit struct and taking a DynamicLimit here instead of the 2 callbacks.

The documentation for the callback functionality can then go in the DynamicLimit doc rather than the WithDynamicLimit doc.
```

But I am adding the table name as an argument to the error callback.